### PR TITLE
New merge algorithm

### LINKED ISSUE #65: Add comments to the example implementation
The example implementation requires comments as it is the example.

### PR DESCRIPTION
**Pull request by koresar**
The PR looks large, although there is only one actual change - the new merging algorithm. It implements #101 

Minor changes:
* update linting
* add plentiful of comments. Closes #65 at last
* Uses new `check-compose` to test the algorithm

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -1,187 +1,9 @@
...
-    "spaced-line-comment": [0, "always"],
-    "strict": [2, "never"],
-    "use-isnan": [2],
-    "valid-jsdoc": [0],
-    "valid-typeof": [2],
-    "vars-on-top": [0],
-    "wrap-iife": [2],
-    "wrap-regex": [2],
-    "yoda": [2, "never", {"exceptRange": true}]
+    "no-param-reassign": 0,
```
Comment (FredyC) [inline .eslintrc]: Why to disable this rule? I consider that quite an anti-pattern to change something I have received in arguments.

File: package.json [hunk 2]
```diff
@@ -29,12 +29,13 @@
...
     "babel-cli": "^6.6.5",
-    "babel-eslint": "^5.0.0",
+    "babel-eslint": "^6.1.2",
     "babel-polyfill": "^6.7.2",
     "babel-preset-es2015": "^6.6.0",
     "babel-register": "^6.7.2",
-    "check-compose": "^2.0.0",
-    "eslint": "~2.2.0"
+    "check-compose": "^3.1.2",
+    "eslint": "^2.13.1",
```
Comment (FredyC) [inline package.json]: If I understand correctly, you don't want to go with ESlint v3 anytime soon since it has dropped support for Node < 4. I am wondering when it will be right time to drop that support from stampit as well. I think that people who deliberately stay on 0.12 are kinda kicking to their own knees as there is a lot of performance improvements missing not even mentioning native support for a lot of ES6 stuff.

File: compose.js [hunk 3]
```diff
@@ -21,20 +60,29 @@ export const merge = (dst, src) => mergeWith(dst, src, (dstValue, srcValue) => {
...
 
+    // Deep merge, then override with shallow merged properties, then apply property descriptors.
     merge(obj, descriptor.deepProperties);
     assign(obj, descriptor.properties);
     Object.defineProperties(obj, descriptor.propertyDescriptors || {});
 
+    // Run initializers sequentially.
     return (descriptor.initializers || [])
       .filter(isFunction)
       .reduce((resultingObj, initializer) => {
```
Comment (FredyC) [inline compose.js]: Perhaps rename `resultingObj` to `resultingInstance` ?

File: compose.js [hunk 4]
```diff
@@ -21,20 +60,29 @@ export const merge = (dst, src) => mergeWith(dst, src, (dstValue, srcValue) => {
  */
 function createFactory(descriptor) {
   return function Stamp(options = {}, ...args) {
-    let obj = Object.create(descriptor.methods || {});
+    // The 'methods' metadata object becomes the prototype for new object instances.
+    const obj = Object.create(descriptor.methods || {});
```
Comment (FredyC) [inline compose.js]: Could be also named `instance` instead of just generic `obj` to clearly represent what is it about.

**Comment by FredyC** [comment]
Isn't there already some consensus to actually remove code from the specification repo? I believe @ericelliott has agreed on that too. The `stampit` has already side tracked little with separating merging somewhere else. Do you plan to copy&paste this back there?

File: .eslintrc (hunk 1 above)
Comment (boneskull) [inline .eslintrc]: I don't care either way about this, but the modifications to `.eslintrc` seem like they belong in a different PR.

File: README.md [hunk 5]
```diff
@@ -141,31 +141,42 @@ The stamp descriptor properties are made available on each stamp as `stamp.compo
...
+* `properties` are copied by assignment
+* `deepProperties` are deep merged
+* `propertyDescriptors` are copied by assignment
+* `staticProperties` are copied by assignment
+* `staticDeepProperties` are deep merged
+* `staticPropertyDescriptors` are copied by assignment
 * `initializers` are uniquely concatenated as in `_.union()`.
-* `configuration` are copied by assignment as in `Object.assign()`.
-* `deepConfiguration` are deep merged, except arrays are concatenated.
+* `configuration` are copied by assignment
```
Comment (boneskull) [inline README.md]: I don't understand the difference between `configuration` and `properties`, or what drove this.  Can you send me to a discussion or further documentation...?

File: README.md [hunk 6]
```diff
@@ -141,31 +141,42 @@ The stamp descriptor properties are made available on each stamp as `stamp.compo
...
+* `deepConfiguration` are deep merged
 
+##### Copying by assignment
+
+The regular `Object.assign()` is used.
+
+##### Deep merging
+
+Special deep merging algorithm should be used when mergind descriptors:
```
Comment (boneskull) [inline README.md]: typo

File: README.md [hunk 7]
```diff
@@ -141,31 +141,42 @@ The stamp descriptor properties are made available on each stamp as `stamp.compo
...
+* The last object type always overwrite previous object type
```
Comment (boneskull) [inline README.md]: should probably read:

```md
* The last object type always overwrites the previous object type
```

File: README.md [hunk 8]
```diff
@@ -141,31 +141,42 @@ The stamp descriptor properties are made available on each stamp as `stamp.compo
...
+* Plain objects are deeply merged or cloned
```
Comment (boneskull) [inline README.md]: this is a bit confusing.  when are objects merged?  when are objects cloned?

File: README.md [hunk 9]
```diff
@@ -141,31 +141,42 @@ The stamp descriptor properties are made available on each stamp as `stamp.compo
...
+* Arrays are concatenated, elements should not be changed
```
Comment (boneskull) [inline README.md]: I *think* I understand what you're getting at here, but this doesn't make sense to the layperson that doesn't know offhand that calling `myArray.concat()` just gives you a shallow copy.

"Elements should not be changed" is ambiguous as to whether or not the elements are cloned.

Also: what happens to my properties?  What happens to array-like objects?

```js
const arr = [];
arr.foo = 'bar';
```

File: README.md [hunk 10]
```diff
@@ -141,31 +141,42 @@ The stamp descriptor properties are made available on each stamp as `stamp.compo
...
+* Functions, Symbols, RegExp, etc. are copied by assignment
```
Comment (boneskull) [inline README.md]: What about properties of functions?  

"Etc." is likely insufficient; everything that's given special treatment should be enumerated here.

File: compose.js [hunk 11]
```diff
@@ -21,20 +60,29 @@ export const merge = (dst, src) => mergeWith(dst, src, (dstValue, srcValue) => {
...
         const returnedValue = initializer.call(
+          // 'this' context will be the object instance itself
           resultingObj,
+          // the first argument passed from factory to initializer
           options,
+          // special arguments. See specification.
           {instance: resultingObj, stamp: Stamp, args: [options].concat(args)}
         );
+
+        // Any initializer can override the object instance with basically anything.
```
Comment (boneskull) [inline compose.js]: Out of curiosity, will Babel transpile this in a ES3-friendly way?  Perhaps it's safer to use `typeof` here.

File: compose.js [hunk 12]
```diff
@@ -69,40 +121,50 @@ function createStamp(descriptor, composeFunction) {
...
-  const srcDescriptor = (srcComposable && srcComposable.compose) || srcComposable;
+  // Check if it's a stamp or something else.
+  const srcDescriptor = isStamp(srcComposable) ? srcComposable.compose : srcComposable;
+  // Ignore everything but things we can merge.
   if (!isDescriptor(srcDescriptor)) return dstDescriptor;
 
   const combineProperty = (propName, action) => {
-    if (!isObject(srcDescriptor[propName])) return;
-    if (!isObject(dstDescriptor[propName])) dstDescriptor[propName] = {};
+    // Do not create destination properties if there is no need.
```
Comment (boneskull) [inline compose.js]: Out of curiosity, what's the advantage to using a lambda here instead of a function expression or statement?

File: compose.js [hunk 13]
```diff
@@ -69,40 +121,50 @@ function createStamp(descriptor, composeFunction) {
...
-    const initializers = (dstDescriptor.initializers || [])
-      .concat(srcDescriptor.initializers)
-      .filter(isFunction);
-    dstDescriptor.initializers = uniq(initializers);
+  combineProperty('deepConfiguration', mergeDescriptor);
+
+  if (isArray(srcDescriptor.initializers)) {
+    // Initializers must be concatenated. '.concat' will also create a new array instance.
+    const dstInitializers = (dstDescriptor.initializers || []).concat(srcDescriptor.initializers);
+    // The resulting initializers array must contain function only, and
```
Comment (boneskull) [inline compose.js]: `s/function/functions`

File: compose.js [hunk 14]
```diff
@@ -69,40 +121,50 @@ function createStamp(descriptor, composeFunction) {
...
+    // must not have duplicate initializers - the first occurrence wins.
+    dstDescriptor.initializers = uniq(dstInitializers.filter(isFunction));
```
Comment (boneskull) [inline compose.js]: What's driving elimination of duplicate initializers?  To me, this seems like unexpected behavior.

File: compose.js [hunk 15]
```diff
@@ -69,40 +121,50 @@ function createStamp(descriptor, composeFunction) {
...
   }
 
   return dstDescriptor;
 }
 
 /**
- * Given the list of composables (stamp descriptors and stamps) returns a new stamp (composable factory function).
+ * Given the list of composables (stamp descriptors and stamps)
+ * returns a new stamp (composable factory function).
```
Comment (boneskull) [inline compose.js]: If you can make a `@typedef` that defines a Stamp, I'll buy you a beer.

File: compose.js [hunk 16]
```diff
@@ -69,40 +121,50 @@ function createStamp(descriptor, composeFunction) {
...
  * @param {...(object|Function)} [composables] The list of composables.
  * @returns {Function} A new stamp (aka composable factory function).
  */
 export default function compose(...composables) {
-  const descriptor = [this].concat(composables).filter(isObject).reduce(mergeComposable, {});
+  // Merge metadata of all composables to a new plain object.
+  const descriptor = [this].concat(composables).filter(isMergeable).reduce(mergeComposable, {});
+  // Pass self as the 'compose' implementation which will be used for `Stamp.compose()`
```
Comment (boneskull) [inline compose.js]: What's "self" here?  Maybe this could be more clear

File: package.json (hunk 2 above)
Comment (boneskull) [inline package.json]: @FredyC Why does Stampit have to run on Node?

Comment (FredyC) [inline package.json]: @boneskull Well if you write client side code only it's fine, but isomorphic apps are more then required these days.

File: README.md (hunk 10 above)
Comment (FredyC) [inline README.md]: Functions are copied by a reference, see #101

File: README.md (hunk 5 above)
Comment (FredyC) [inline README.md]: Check https://github.com/stampit-org/stamp-specification#the-stamp-descriptor

File: README.md (hunk 9 above)
Comment (FredyC) [inline README.md]: Yea well, with this merging mechanism, array properties would be lost however I find that fairly bad practice. Object should be either list of values or map of values, not both at once.

File: package.json (hunk 2 above)
Comment (boneskull) [inline package.json]: I guess my comment was ambiguous.

When you say "drop support for stamps", what do you mean exactly?

File: compose.js (hunk 11 above)
Comment (FredyC) [inline compose.js]: What environments are running ES3 these days? Perhaps unsupported IE 8? I don't see reason why to support something that is abandonded in general.

File: compose.js (hunk 12 above)
Comment (FredyC) [inline compose.js]: Probably just readability, no added value in this for sure, but also no harm.

File: compose.js (hunk 14 above)
Comment (FredyC) [inline compose.js]: Check #52

File: README.md (hunk 9 above)
Comment (boneskull) [inline README.md]: In general, I'm pro-"interoperability with ES6 classes", though that discussion has been had elsewhere.  I don't feel like Stampit should be an all-or-nothing proposition.  Currently, I can't use a `Map` or `Set` in my `props` object.  I wouldn't be able to use some weird subclass of `Array` either (though I wouldn't write one).

File: README.md (hunk 10 above)
Comment (boneskull) [inline README.md]: Is the comment incorrect, then?

File: compose.js (hunk 15 above)
Comment (FredyC) [inline compose.js]: Check `isStamp` in the top, that's all there is to it. It has to be a function and have a `compose` function, nothing else.

File: compose.js (hunk 11 above)
Comment (boneskull) [inline compose.js]: IE7/IE8, yes.  I'm with you there, but this doesn't address my question.

File: compose.js (hunk 12 above)
Comment (boneskull) [inline compose.js]: (Used to be that you wanted to name your functions for better stack traces, but I think there's been improvement there recently)

File: package.json (hunk 2 above)
Comment (FredyC) [inline package.json]: I mean to officially drop support for Node < 4 in a stamp code. It doesn't necessarilly mean any code changes, currently all tests are passing down to 0.10. However eslint v3 has dropped support for these old Node versions so tests would be failing for these.

File: compose.js (hunk 11 above)
Comment (FredyC) [inline compose.js]: I believe it does :) Babel transpiles code to ES5 not to ES3.

File: README.md (hunk 10 above)
Comment (FredyC) [inline README.md]: No it's correct, assigment is essentially copying by reference. It wont change objects of those types in any way.

File: README.md (hunk 9 above)
Comment (FredyC) [inline README.md]: I am lost. How is this relevant to ES6 classes? Classes don't do any array concatenation nor merging or assigning.

File: compose.js (hunk 12 above)
Comment (FredyC) [inline compose.js]: Well Babel transpiles this essentially to `var combineProperty = function combineProperty(...` so you don't need to worry about that really.

File: README.md (hunk 9 above)
Comment (boneskull) [inline README.md]: You can subclass `Array`, producing an array-like object, which may have properties, which would be lost upon a `concat()`, assuming the array-like object didn't override `Array.prototype.concat()`.  

Yes, detestable, I know.  But in my perfect world, I would be able to take any pseudoclass, ES6 class or subclass thereof, and coerce it into a Stamp, clone it as a property, or deeply merge it as applicable.

Comment (FredyC) [inline README.md]: I had never need to subclass Array, I find that rather confusing really. If anything, I would rather make a proxy class which hold array instance and offers necessary methods, but that's not the point here. It seems strange to me that to combine two such different paradigms in a single codebase. I don't think we should support that really as our goal here is to show that classes are bad thing and people shouldn't be using those.

Comment (ericelliott) [inline README.md]: >  I wouldn't be able to use some weird subclass of Array either (though I wouldn't write one).

It is not currently safe to subclass array if you want cross-browser support, and probably won't be for at least a year.

File: .eslintrc (hunk 1 above)
Comment (koresar) [inline .eslintrc]: @boneskull you are correct. I have these changes in my git for quite a while. It was just quicker way for me.

@FredyC it is because of `mergeComposable` function. Namely these two lines:
```js
if (!isDescriptor(dstDescriptor[propName])) dstDescriptor[propName] = {};
```
and
```js
dstDescriptor.initializers = uniq(dstInitializers.filter(isFunction)
```

File: package.json (hunk 2 above)
Comment (koresar) [inline package.json]: I believe this repo can drop support for anything. This is just a sample implementation. We do not care of it cross compatibility.

I can migrate this PR to use the latest eslint if you insist.

I put eslint changes to this PR because I had them laying on my hdd for some time. It was simpler that way. That's why it is eslint@2. Surely I can bump it to eslint@3. Should I do it in this PR?

File: compose.js (hunk 3 above)
Comment (koresar) [inline compose.js]: done

File: compose.js (hunk 4 above)
Comment (koresar) [inline compose.js]: done

**Comment by koresar** [comment]
@FredyC 
> Do you plan to copy&paste this back there?

Yes. I already did it in my local git repo.

File: README.md (hunk 5 above)
Comment (koresar) [inline README.md]: Thanks @FredyC for the link.
TL;DR:
* properties are assigned to instances,
* configuration is not, but can be effectively used in initializers or static functions for various scenarios. See Fun With Stamps [6](https://medium.com/@koresar/fun-with-stamps-episode-6-statics-properties-on-stamps-e2b85825152c) and [7](https://medium.com/@koresar/fun-with-stamps-episode-8-tracking-and-overriding-composition-573aa85ba622) for more examples.

File: README.md (hunk 6 above)
Comment (koresar) [inline README.md]: fixed. thanks!

File: .eslintrc (hunk 1 above)
Comment (FredyC) [inline .eslintrc]: I see. Perhaps you might just disable that rule in a code instead of global disable because sometimes it can discover bad practice.

File: README.md (hunk 9 above)
Comment (koresar) [inline README.md]: @boneskull there are few other ways to merge `Map` or `Set` in stamps.
Via [infected compose](https://medium.com/@koresar/fun-with-stamps-episode-8-tracking-and-overriding-composition-573aa85ba622) or static methods. E.g.:
```js
const AnotherStamp = Stamp.mergeMyMap(...).mergeMySetPlease(...);
```

If you have more questions about the merging algorithm please leave your comments in #101. We'd be glad to discuss and modify the spec. as needed.

(Looks like you've been gone for too long.)

File: README.md (hunk 10 above)
Comment (koresar) [inline README.md]: Changing it to `are copied by reference`. It's a bit less confusing. Thanks @boneskull

File: compose.js (hunk 11 above)
Comment (koresar) [inline compose.js]: Yes, babel is ES5 transpiler. Moreover, in this repo we do not about any older environment support because this is just a reference implementation of the `compose` function.

File: compose.js (hunk 12 above)
Comment (koresar) [inline compose.js]: This main goal of the code is to be highly readable. If you find `function combineProperty()` more readable than the lambda then I'm happy to rewrite that line.

File: compose.js (hunk 13 above)
Comment (koresar) [inline compose.js]: oh! such a good reviewer you are! thanks mate 👍

**Comment by FredyC** [comment]
@koresar Ok but still I don't see a reason why to keep the code in here. Check the comment here: https://github.com/stampit-org/stampit/issues/221#issuecomment-231564585

File: compose.js (hunk 15 above)
Comment (koresar) [inline compose.js]: @boneskull
```js
/**
 * @typedef {Function} Stamp
 * @property {Function} compose
 */
```
So, where to get my beer? 😆 

But on the more serious note, give me a minute.

Comment (koresar) [inline compose.js]: I'm adding this to the top of the file:
```js
/**
 * @typedef {Function} Stamp
 * @property {Function} compose
 * @property {Object} [compose.methods] Instance ptototype methods
 * @property {Object} [compose.properties] Instance properties
 * @property {Object} [compose.deepProperties] Instance deep merged properties
 * @property {Object} [compose.propertyDescriptors] JavaScript property descriptors
 * @property {Object} [compose.staticProperties] Stamp static properties
 * @property {Object} [compose.staticDeepProperties] Stamp deep merged static properties
 * @property {Object} [compose.staticPropertyDescriptors] Stamp JavaScript property descriptors
 * @property {Object} [compose.configuration] Stamp configuration
 * @property {Object} [compose.deepConfiguration] Stamp deep merged configuration
 */
```

File: compose.js (hunk 16 above)
Comment (koresar) [inline compose.js]: Changing to:
> // Recursively pass this 'compose' implementation which will be used for `Stamp.compose()`

**Comment by koresar** [comment]
@FredyC I'm not seeing what you mean. This code is an **example implementation**.

Someday other people would like to make their own implementation of stamps. This repo will give them more understanding of what, and why, and how. That's basically it.

File: .eslintrc (hunk 1 above)
Comment (koresar) [inline .eslintrc]: This is not a production code! We don't need to discover any bad practices.
Please, do not forget, the `compose.js` must be highly readable. I would like to avoid adding any linting PRAGMA codes whatsoever. :)

File: package.json (hunk 2 above)
Comment (koresar) [inline package.json]: @boneskull this is not `stampit` repo. But `stamp-specification`. :)

**Comment by koresar** [comment]
I'm merging this in few hours unless someone have anything to say.

**Comment by FredyC** [comment]
Yea, I'll save you waiting, I think it looks good, it just feels weird to have any code in here. Eric wasn't against the idea of leaving only specification here, but if you want to maintain it here as well, it's your choice.

**Comment by koresar** [comment]
@FredyC What's your proposal? Where the example implementation better reside?

**Comment by FredyC** [comment]
I don't think there needs to be any example implementation really, especially after that most performant `compose` is made, I don't see reason why would anyone wrote things differently and yet conform to spec.

**Comment by koresar** [comment]
True. But let's keep this example implementation here for a while. I go to it myself from time to time. Mostly when I do presentation(s) about stamp on various meetups/conferences and people are asking questions. So, I just show them that file. I need it. :)

File: README.md (hunk 9 above)
Comment (boneskull) [inline README.md]: > It seems strange to me to combine two such different paradigms in a single codebase.

Mostly I'm looking at interoperability between my codebase and somebody else's codebase.  A subclassed `Array` was a contrived example.  

More likely, someone will pick up Stampit and wonder why they have to write custom wrappers to be able to conveniently use a built-in like `Set`.

I also understand that I'm in opposition to the general consensus 'round here.  I want to discourage the use of ES6 classes just as much as anyone else here.  *How* the project builds towards that goal is where I disagree.  

But about all I can do is tell you about it.  If anyone is tiring of my opinion, I'll stfu.  :smile:

Comment (koresar) [inline README.md]: Hey, we like all the ideas!
If you could present the idea about handling the `Set` as a GitHub issue, or better as a PR, that would be great beginning towards a productive discussion.
